- use major-only for Clang >= 8.0

SSE4 opened the pull request:
according to the:
https://releases.llvm.org/7.0.0/tools/clang/docs/ReleaseNotes.html
```
Clang binary and libraries have been renamed from 7.0 to 7. For example, the clang binary will be called clang-7 instead of clang-7.0.
```
also according to the:
http://lists.llvm.org/pipermail/llvm-dev/2018-March/121873.html
```
Now, as we are not using the minor version (the Y in X.Y.Z), there isn't
much point in calling our tools foo-7.0 as we won't create a 7.1.
```
and finally:
http://blog.llvm.org/2016/12/llvms-new-versioning-scheme.html
```
We decided that going forward, every release on the six-month cycle will be a major release. Patch releases will increment the patch component as before (producing versions like 5.0.1), and the minor component will stay at zero since no minor releases will be made.
```
so, this patch offers to use major only versions for Clang >= 7.0, similarly on how it is done for GCC >= 5.0

- [ ] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. Also adding a description of the changes in the ``changelog.rst`` file. https://github.com/conan-io/docs

SSE4 commented:
build problem appears to be an infrastructure issue:
```
ERROR: Issue with creating launcher for agent MacosBuilder. The agent has not been fully initialized yet
```

memsharded commented on conans/client/conf/detect.py:
```diff
@@ -64,10 +64,15 @@ def _clang_compiler(output, compiler_exe="clang"):
         installed_version = re.search("([0-9]+\.[0-9])", out).group()
         if installed_version:
             output.success("Found %s %s" % (compiler, installed_version))
+            major = installed_version.split(".")[0]
```
I am afraid this would be a breaking change, what do you think @lasote?

Croydon replied:
It's only for Clang and version >= 7.
What do you fear that this would break?

lasote replied:
If you are trusting your default profile and you have packages generated for "7.0" It won't find the "7" packages anymore.

SSE4 replied:
how does it work for GCC 5.0 vs GCC 5 for instance?

SSE4 replied:
at least I didn't spot any GCC-specifc quirks in ConanInfo implementation, could you point me?

Croydon replied:
What about changing it only for Clang >= 8 then and allowing via default Conan setting.yml both values 7 and 7.0 for the moment?

This could be left this way or changed retrospectively with Conan 2.0

Since Clang 8 is not releases yet it wouldn't break existing behaviour.

SSE4 replied:
I agree with Croydon
how about changing it for Clang >= 8.0, and use docker images `conan/lasoteclang70` but `conan/lasoteclang8`?

SSE4 replied:
after the today's discussion, we decided:
- [x] keep Clang 7.0 to avoid compatibility breakage
- [x] since Clang 8, use major only
- [x] for docker images, use `conanio/clang70` and `conanio/clang8`, for the consistency
pull request has been updated accordingly, see notes below

SSE4 commented on conans/client/migrations.py:
```diff
@@ -94,7 +94,8 @@ def _make_migrations(self, old_version):
         version: ["8", "9", "10", "11", "12", "14", "15"]
         toolset: [None, v90, v100, v110, v110_xp, v120, v120_xp, v140, v140_xp, v140_clang_c2, LLVM-vs2014, LLVM-vs2014_xp, v141, v141_xp, v141_clang_c2]
     clang:
-        version: ["3.3", "3.4", "3.5", "3.6", "3.7", "3.8", "3.9", "4.0", "5.0", "6.0", "7.0"]
```
okay

SSE4 replied:
reverted it back

SSE4 commented:
same question, do we have such quirks for GCC? also, how about patch versions for GCC? I guess they currently don't matter, assuming all patch versions are compatible, if major is same, right?
so, I wonder, why Clang logic must be different from GCC's? or do you mean it should be changed for both GCC and Clang?

SSE4 commented:
updated, summary:
- [x] added Clang 8 to settings.yml
- [x] starting from Clang 8, conan will detect major-only by default
- [x] detection for Clang 7 and below is not changed to avoid compatibility breakage
- [x] added major-only Clang versions 5, 6, 7, 8 to settings.yml
- [x] migrations.py is not changed
- [x] added missing compilers to the new_ci.py (GCC 8, Clang 5, 6, 7, Apple Clang 10.0)
- [x] CMake-wrapper checks major-only since Clang 8 as well

Croydon commented on conans/client/conf/__init__.py:
```diff
@@ -67,7 +67,11 @@
+                  "5", "5.0",
+                  "6", "6.0",
+                  "7", "7.0",
+                  "8", "8.0"]
```
If we want to change it starting with 8, why did you still add `8.0`?

lasote commented:
@memsharded please confirm that we are following this approach of keeping `clang 7.0` and merge when you are ready.
